Add driver for Smile T, some extra capabilities for Anna and Adam HA devices, refactor of library/code

synegic opened the pull request:
This PR adds some of the missing functionality the new app lacked for users that had the Anna app. It includes a simple 'Set scene' Flowcard action, measure_luminance, alarm_motion for Anna and measure_pressure and measure_temperature.outside for the bridge devices.

I could only test the changes with the Anna + Smile T pair, as I don't have any other Plugwise devices. I was not entirely sure how I'd go about setting location_preset for Adam, as to me Adam seems to be the gateway device and not the 'location preset' device linked up with a 'zone', so I left the Flowcard action out for Adam.

We might be able to combine both Adam and Smile into one single 'Plugwise bridge'-type of device, as they currently both use the same appliance exposed by the Plugwise API as "Central heating boiler". They are more or less be the same capability/flow-wise.

Currently it's required to add both devices (Anna+Smile) separately, not sure what the possibilities are to add them both at the same time in the current set-up.

Would be great if someone with other Plugwise devices could take a look if these changes work for them. Let me know what you think.

I aim add more Flow actions and conditions over the next few weeks to come.

TedTolboom commented:
Nice work @synegic !
I'll try to free-up some time this week to test on my Anna and Adam

bckp commented:
Fixes #5

synegic commented:
So, I've just added another commit which should improve a lot of things, in addition to initial pull request. I've attempted to increase chances of devices being found by discovery (as this sometimes was a problem), refactored the lib and driver classes to keep code duplication to a minimum, added some templates for homeycompose for the flows (to be extended later on). It also adds boiler_error_code:changed and location_preset:changed trigger flowcards to their supported devices, with accompanying tags, which should make this app a bare minimum replacement of the original Anna app.

Once again; I only can test for Anna/Smile T combination and would really love to hear feedback from someone who owns more Plugwise stuff to test. :-)

bckp commented on .homeycompose/capabilities/boiler_state.json:
```diff
@@ -1,22 +1,30 @@
-		}
-	},
+    "true": {
+      "en": "Flame on",
```
This is just cosmetic, by my electric heater does not have flame :-D isn't better to call it heater, and state should be ACTIVE and IDLE?

TedTolboom commented on .homeycompose/capabilities/DHW_state.json:
```diff
@@ -1,22 +0,0 @@
-{
```
Changing the name of a custom capability will break functionality for existing users: suggest to keep `DHW_state` as ID

TedTolboom commented on .homeycompose/capabilities/dhw_state.json:
```diff
@@ -0,0 +1,30 @@
+{
```
Changing the name of a custom capability will break functionality for existing users: suggest to keep DHW_state as ID

TedTolboom commented on .homeycompose/capabilities/location_preset.json:
```diff
@@ -38,7 +38,7 @@
       "id": "no_frost",
       "title": {
-        "en": "Frost protection",
+        "en": "Anti-freeze",
```
I'd use "Frost protection" and "Vorstbeveiliging", but that is a personal taste... Anti-freeze I put into my car and doesn't cover the principle

synegic replied:
I hope you don't actually personally taste the antifreeze you put in your car. ;-)

Was a valid suggestion. Changed.

TedTolboom commented on .homeycompose/drivers/templates/flow.gateway.json:
```diff
@@ -0,0 +1,57 @@
+  "$flow": {
+    "actions": [
+      {
+        "id": "{{driverId}}:dhw_mode:set",
```
This is causing to crash my existing setup, since the existing app is using `Set_DHW_mode` as ID.
https://github.com/athombv/com.plugwise.adam/blob/master/drivers/adam/driver.flow.compose.json#L6
My suggestion to resolve: move definition of this flow card (original driver.flow.compose.json) to `.homeycompose/flow/actions/Set_DHW_mode.json` and use addition device argument to only make it available based on driver ID (Adam and Smile). In this way, it will not break the existing functionality

synegic replied:
I've renamed the driverId: parts to their old counterparts. So they should fit now and thus not incur a crash anymore.

TedTolboom commented on .homeycompose/drivers/templates/flow.gateway.json:
```diff
@@ -0,0 +1,57 @@
+    ],
+    "triggers": [
+      {
+        "id": "{{driverId}}:boiler_error_code:changed",
```
New card, so should work / not break existing functionality.
But suggestion could be to keep similar approach to set_DHW_mode

TedTolboom commented on .homeycompose/drivers/templates/flow.thermostat.json:
```diff
@@ -0,0 +1,79 @@
+  "$flow": {
+    "actions": [
+      {
+        "id": "{{driverId}}:location_preset:set",
```
Same breaking functionality as mentioned for set_DHW_mode. Please change.

synegic replied:
As far as I can tell, there is no such flowcard yet. So this shouldn't break anything. I've changed the naming nonetheless, to be in line with the Set_DHW_Mode one.

TedTolboom commented on drivers/adam/driver.compose.json:
```diff
@@ -1,40 +1,48 @@
   "capabilities": [
     "boiler_state",
-    "DHW_state",
+    "dhw_state",
```
don't change capability names; will break functionality for Adam users

TedTolboom commented on drivers/adam/driver.compose.json:
```diff
@@ -1,40 +1,48 @@
-    "measure_temperature.intended",
-    "measure_temperature.boiler",
+    "measure_pressure",
+    "measure_temperature.target",
```
don't change capability names; will break functionality for Adam users

TedTolboom commented on drivers/adam/driver.compose.json:
```diff
@@ -1,40 +1,48 @@
-    "measure_temperature.boiler",
+    "measure_pressure",
+    "measure_temperature.target",
+    "measure_temperature.output",
```
don't change capability names; will break functionality for Adam users

TedTolboom commented on drivers/adam/driver.compose.json:
```diff
@@ -1,40 +1,48 @@
   ],
   "capabilitiesOptions": {
-    "measure_temperature.intended": {
+    "measure_temperature.target": {
```
don't change capability names; will break functionality for Adam users

TedTolboom commented on drivers/adam/driver.compose.json:
```diff
@@ -1,40 +1,48 @@
+      }
     },
-    "measure_temperature.boiler": {
+    "measure_temperature.output": {
```
don't change capability names; will break functionality for Adam users

synegic commented:
Thanks for the review @TedTolboom, much appreciated. I've done another commit to resolve most of the backward incompatibilities. Let me know how it looks (when you have time, of course!).

TedTolboom commented on .homeycompose/capabilities/dhw_state.json:
```diff
@@ -26,5 +26,5 @@
   },
   "uiComponent": "sensor",
-  "icon": "/assets/capabilities/dhw_state.svg"
+  "icon": "/assets/capabilities/DHW_state.svg"
```
Check validation: you'll need to change the actual filename of the vector to `DHW_state.svg`

TedTolboom commented on app.json:
```diff
@@ -395,25 +395,25 @@
         "boiler_state",
-        "dhw_state",
-        "dhw_mode",
+        "DHW_state",
```
Check validation: need to rename /capabilities/dhw_state.json to DHW_state.json

TedTolboom commented on app.json:
```diff
@@ -395,25 +395,25 @@
-        "dhw_state",
-        "dhw_mode",
+        "DHW_state",
+        "DHW_mode",
```
Check validation: need to rename /capabilities/dhw_mode.json to DHW_mode.json

synegic commented:
> Check the validation; some filenames still need to be updated

Weird, apparently my OS FS is case insensitive and thus GitHub Desktop does not seem to pick up these changes..

TedTolboom commented:
Validating now correctly... will start testing functionality
DHW mode on Adam is working correctly (not breaking)

TedTolboom commented:
`outdoor_temperature` doesn't work on both smile and Adam. 
You're depending on the `appliance` polling, but `outdoor_temperature` is reported from within locations.

Also pressure not, but that one is not visible in /core/appliances so that might be my boiler not reporting it.

synegic commented:
> `outdoor_temperature` doesn't work on both smile and Adam.
> You're depending on the `appliance` polling, but `outdoor_temperature` is reported from within locations.
Hmm, I suppose that the location outdoor_temp is actually the temperature reported by the 'weather service' functionality Plugwise offers. In my case, outdoor temp is actually registered by the boiler as I have an outdoor temp sensor. So I suppose, because we actually want sensor values there, it is correct but just not reported if you do not have such sensor (which is correct).

> Also pressure not, but that one is not visible in /core/appliances so that might be my boiler not reporting it.
Probably for the same reason as outdoor temp is not working indeed; not supported by the boiler.

Is it possible to hide/add/enable/disable capabilities in a dynamic way?

bckp commented:
> So I suppose, because we actually want sensor values there, it is correct but just not reported if you do not have such sensor.
Isn`t better to (if you have outdoor temp sensor add data from it, and if not, get the wheater location data from plugwise service?)

TedTolboom commented:
yes, this is possible... didn't use it myself yet
among others the Enelogic app is using this approach to differentiate between electrical and water based capabilities.

synegic commented:
Can't really seem to find any documentation or references on that. Maybe you could link me to it? I've tried to see if bckp's suggestion would work, but that wouldn't work without doing an extra call, including location listeners for all gateway devices and exposing that data to appliance's onPoll as well, which I'd like to avoid for this one thing.

TedTolboom commented:
@synegic let's discuss that question on Slack #developers

TedTolboom commented on lib/PlugwiseGatewayDevice.js:
```diff
@@ -0,0 +1,165 @@
+
+  async onCapabilityDHWmode(value) {
+    return this.bridge.setDHWmode({
+      applianceId: this.applianceId,
```
`this.applianceID` does not exist here, and therefore the action card is not working.

best option could be to get the applianceId out of the arguments returned from the flow card; `args.device`

synegic replied:
Are you sure? As we bind the call to the onCapabilityDHWMode to 'this' in the onInit?

TedTolboom replied:
yes, I'm sure.
Try yourself to set the DHW mode... flow card will show error: applianceId not defined

synegic replied:
![Screenshot 2019-05-31 at 10 23 01](https://user-images.githubusercontent.com/30956886/58692122-389e3d80-838e-11e9-8b78-90e5e4d90261.png)
Works here.. As far as I can tell, this method of using applianceId there is used at almost all places. From the base PlugwiseDevice class onInit it is actually fetched by getData. But I will have a look if I could get it from args.device.

TedTolboom replied:
Check if it actually changes the setting... http://IP/core/appliances

At my setup, it is using the wrong applianceID... so yes, green checkmark, but setting is not changed

synegic replied:
`<type>domestic_hot_water_comfort_mode</type>
<state>off</state>`

Also changes the setting without any problems. But I will use args.device for flowcard actions, just to be sure. For other operations, this is impossible I guess as there is no other source to get this info than from base class.

synegic replied:
By the way, for me the acuator with type domestic_hot_water_comfort_mode is listed under the appliance with type "<name>Central heating boiler</name>". Is this also the case in your setup?

TedTolboom replied:
yes, under the appliance with name: `<name>Central heating boiler</name>`

synegic replied:
Okay, well, the newest commit should fix it if it's related to the applianceId not available at the mentioned line in code. If not, then something else is wrong. Maybe it has something to do with previous pairing that was done without including applianceId, or maybe even wrong applianceId into the device data?

TedTolboom replied:
Strange, removed the Adam device and now it is working fine...
Still something fishy going on threatening upgradeability

synegic replied:
Then maybe it's a good idea to build in a check on the Driver's onInit to verify applianceId is actually set and/or is still correct, and if not, update that in the device's data stored in Homey? But even if that was the case, if the applianceId was incorrect you wouldn't be seeing any data coming in.. Hmm. Pretty weird issue indeed.

synegic replied:
But if we look at the original, previous version of these files/implementation, the same way is used; https://github.com/athombv/com.plugwise.adam/blob/master/drivers/adam/device.js. So even if this is a bug somewhere, it shouldn't really matter form an upgradeability-point of view as old code was already using it there.

synegic commented:
With latest commit, the Adam/Smile devices should only add the capabilities they actually report as measurements upon pairing.

TedTolboom commented:
> With latest commit, the Adam/Smile devices should only add the capabilities they actually report as measurements upon pairing.

Nice!

TedTolboom commented:
@synegic when I opened my bottle of beer the cause hit me...
You’re combining functionality in the GatewayDevice. In my setup, combining the Adam and Anna gateways, this device instance is used for both gateway devices...

When I re-installed the Smile T central heating device yesterday, the initialization must have set the applianceId for the Smile T.... 

when I tried the set DHW mode flowcard, I used the one of the Adam.. which didn’t work

After re-installing the Adam central heating device, it did work...

Didn’t reproduce yet (beer comes first, after sharing this), but I guess now the flowcard of the Smile T will not work...

synegic commented:
> @synegic when I opened my bottle of beer the cause hit me...
> You’re combining functionality in the GatewayDevice. In my setup, combining the Adam and Anna gateways, this device instance is used for both gateway devices...

Hmm, how would that work? As we are exporting two distinct classes (which indeed are inherited from the GatewayDevice parent class), and they are inited from their own deviceClass, I don't see how that could actually mess things up. Though, I also have no idea what could actually be happening instead..

synegic commented:
@WeeJeWel could you take a look at this? If there's anything left to be changed before it can me merged, please let me know. Thanks!

bckp commented:
ping @WeeJeWel

nlrb commented:
What is the reason this PR is still open? This code brings so much more functionality than the current app code. Great work @synegic!

dennisbrouwer91 commented:
Any update ?

Drardollan commented:
Can anybody please have a look at this update an merge it? We are waiting for months already! Thanks.

WeeJeWel commented:
Hey guys! Nothing has really happened here because at this moment the PR has conflicts. It would also help a great deal if the PR was split into multiple PRs because we cannot merge all proposed changes. In the end we have to support this code, which means we have to be very selective what we allow into the codebase. I hope you understand our business side of things :)

Drardollan commented:
I would love to, but I am not a developer :-(

Sebas- commented:
Thanks for this build @synegic, bought a plugwise anna and plugwise p1 yesterday but didn't expect the homey support was this bad (since I read about good support, but that was for a previous version of the app), after some searching I found this and installed it without problems.

For anyone that finds this, just follow the steps on https://community.athom.com/t/how-to-cli-install-method/198 using the zipped git repo from https://github.com/synegic/com.plugwise.adam

Now I can control my Anna by using Alexa (set temperature to x degrees) and using it in my flows including the motion alarm and lux properties.

The pessimistic developer in me thinks that this PR will probably not get merged soon since the request for splitting up the code is probably not an easy one and it takes time (and it probably works for synegic so realistically he has to do nothing else).

bckp commented:
> Hey guys! Nothing has really happened here because at this moment the PR has conflicts. It would also help a great deal if the PR was split into multiple PRs because we cannot merge all proposed changes. In the end we have to support this code, which means we have to be very selective what we allow into the codebase. I hope you understand our business side of things :)

Sorry Emile, i was supporting you for long time, but this MR have so many conflicts due to fact, you ignore it for almost year! The code here works so much better, was tested by lot of users, even @TedTolboom see it and test it, yet no support, so far i feel like you are not want help of community anymore 👎 that is sad, as Homey have great oportunity to be the BEST of BESTS!

Drardollan commented:
1,5 year and STILL broken. Shame on you Athom!!

WeeJeWel commented:
Closing this PR because the conflicts have never been resolved by the author.
